### ISSUE TITLE
Address Verification implementation

### ISSUE THREAD

**Issue by george0123** [SalesforceFoundation/Cumulus#1739]
Hi,

I have a question - Can I implement address verification logic into the code for Households? Is this possible and if yes, what are the steps that I need to follow?

**Comment by davidhabib** [comment]
please post questions on the NPSP group in the Hub, rather than in gitHub.  If you are asking about implementing an additional address verification service, yes, we support that.  If you are asking about changing our behavior of how we do Address Management, that would require changes to the core code in the package, which isn't modifiable by organizations.

**Comment by george0123** [comment]
Thank you very much for your answer!
I moved my query to the NPSP group.

However if you could also help, I will really appreciate it. We have looked
into the source code and looks like we need to create a class which
implements the ADDR_IValidator interface.
It also seems that the address verification provider is chosen based on
settings from the Address Verification Settings object which is Depricated
in the latest NSPS package. Are the new settings which need to be used to
ensure future compatibility of our plugin?

Thank you in advance

**Comment by ceiroa** [comment]
Yes, your class needs to implement ADDR_IValidator. It also needs to start with "ADDR_". If you look at https://github.com/SalesforceFoundation/Cumulus/blob/dev/src/classes/STG_PanelADDRVerification_CTRL.cls#L187-L192 you'll see how your class needs to named to be displayed in the settings page. Then whatever implementation you choose there will be stored in the settings and be used until you change it.

**Comment by davidhabib** [comment]
note that we haven't yet made ADDR_IValidator global.  it's tracked on this issue: #950

**Comment by ceiroa** [comment]
Great point, Dave. I forgot about that. There are issues with address validation, but they are actually coming from the third-party services. I think the interface is ready (or as ready as it's ever going to be). Let's discuss making it global in the scrum.

**Comment by george0123** [comment]
Thank you for your replies.
We will start working on the address validation implementation on our end.
Could you please let me know when can we expect the interface to be made
global?

**Comment by ceiroa** [comment]
Could you please post here some additional info on the integration you are
working on? Do you plan to make this available to all NPSP users? If so,
which service are you planning on integrating with?

**Comment by george0123** [comment]
Hi Carlos,

We would like to integrate address verifcation from Experian Data Quality
and make it available to other users.
We would like to integrate the below workflows:
1. Address lookup - when you try to enter an address a separate window pops
up and help you quickly find the address you are searching for.
2. Address verification with user interaction - when you enter you address
and you click “save" or “verify” experian data quality virifies your
address and if it is incorrect it provides you with suggestions helping you
to correct it.
3. Address verification without user interaction - verifies your address
without user interaction and marks it as verified or not verified.
Experian Data Quality also supports batch.

I still have not yet been able to test the current address verification
workflow but looking at the code it seems that it is more a silent way of
verifying an addres i.e. point 3 from the above.
Is that the case? And if that is the case I can achieve address
verification by using an after update and after insert trigger. What will
be the difference in the address verification workflow built into the
package and with a trigger?
Will l be able to achieve 1 and 2 above by implementing the ADDR_IValidator
 interface?

Thanks again!

**Comment by judisohn** [comment]
@george0123 How does the integration you're building differ from the Salesforce integration that Experian currently offers? https://appexchange.salesforce.com/listingDetail?listingId=a0N300000055e86EAA

**Comment by george0123** [comment]
Sorry about the delayed response. We are looking for ways to extend the
package to support NSPS. Currently the package cannot plugin to Household
address edit and new address creation UI and it only works asynchronously
with an after insert and after update trigger.

**Comment by george0123** [comment]
We would like to integrate it in a way so that the customer can click on a
button which invokes Experian Data Quality's "Find Address" popup which
helps users quickly find an address and pastes the address back.
If that is not possible we would like to plugin so that when the user
clicks on Save the EDQ "Interactive" workflow which helps the user enter
missing address information (if there is missing address information) kicks
in and verifies the address.
Basically we want to plugin a visual address verification workflow rather
than just provide back end address verification which just marks the
address as verified or not verified.
Is that possible? What is the current address verification workflow for
NPSP?

**Comment by davidhabib** [comment]
The Address verification support in NPSP is currently only for doing the validation asynchronously.  It isn't designed for plug in services to provide their own user interface for interacting with the user while they are entering the data.  That is probably way beyond the scope NPSP will be able to handle anytime soon, since our user's can enter their address from 3 different places (standard contact page, standard account page, and the address object page).  Seems like you might want to look into Lightning components, and how you could create a component that user's could put on their page layouts in Lightning Experience to provide this type of custom ui.

**Comment by george0123** [comment]
Thank you for your assistance!

As you advise Lighting components is something we are researching and
moving forward with our Contact Data Quality solution.
However I am not sure how will a Lightning component be able to get
integrated into the NSPS custom pages. Are you also planing to move your
solution to Lightning components so that customers can created their own
pages with the Lightning App Builder?

As for our current solution we already offer asynchronous address
verification which is invoked via an AIAU trigger which we will offer to
our mutual customer.
Please let me know if implementing the address verification plugging offers
more advanced functionality than implementing the address verification in a
trigger?

**Comment by davidhabib** [comment]
NPSP's address object uses a standard salesforce page for viewing and editing it, so in lightning experience, we should get the lightning ui for free for the address's detail page, and I assume that custom lightning components will be available for any object's page layout.

as for the advantage of our verification mechanism vs doing it yourself in a trigger, I'll leave that for @ceiroa  to comment on.

**Comment by ceiroa** [comment]
NPSP's address verification is also run from a trigger. The main advantage of creating a class that implements the ADDR_IValidator (the same way that ADDR_Cicero_Validator, ADDR_SmartyStreets_Validator, and ADDR_GoogleGeoAPI_Validator do) is that the option gets displayed in the Address section of the settings page. If the class offers batch address verification, the user also see the "Verify all Addresses" button in that same page, and our ADDR_Validator_Batch will take care of the batch process.

**Comment by njjc** [comment]
@george0123 Check out SmartyStreets' new integration, it does inline address suggestion and completion: https://smartystreets.com/docs/salesforce
